tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -0,0 +1,261 @@
+                    return;
+                }
+                ProcessBuilder processBuilder = new ProcessBuilder(spec.command());
+                process = processBuilder.start();
+                Future updater = executor.submit(new StatusUpdater());
+                Future errorUpdater = executor.submit(new ErrorUpdater());
+                controlChannel = new PrintWriter(new OutputStreamWriter(process.getOutputStream(), StandardCharsets.UTF_8), true);
+                new ControlCommand("start", spec).issue();
+                log.info("{}: Started the process {} to execute command: {}", id, process.toString(), spec.command());
+                process.waitFor();
+                updater.get();
+                errorUpdater.get();
+                int workerExitValue = process.exitValue();
+                log.info("{}: The process finished with the exit value: {}.", id, workerExitValue);
+                if (workerExitValue != 0 && errMsg.isEmpty()) {
+                    errMsg = "ExternalWorker exited with error code " + process.exitValue();
+                }
+                log.info("{}: StatusUpdater and errorUpdater are terminated.", id);
+                doneFuture.complete(errMsg);
+            } catch (IOException e) {
+                log.info("{}: Stdout of the process is closed.", id);
+            } catch (InterruptedException e) {
+                log.info("{}: ProcessMonitor is interrupted.", id);
+            } catch (Exception e) {
+                WorkerUtils.abort(log, "ProcessMonitor", e, doneFuture);
+            } finally {
+                if (process != null && process.isAlive()) {
+                    process.destroy();
+                }
+            }
+        }
+    }
+
+    public class ControlCommand {
+        private final String action;
+        private final ExternalCommandSpec spec;
+        @JsonCreator
+        ControlCommand(@JsonProperty("action") String action,
+                       @JsonProperty("spec") ExternalCommandSpec spec) {
+            this.action = action;
+            this.spec = spec;
+        }
+        @JsonProperty
+        public String action() {
+            return action;
+        }
+        @JsonProperty
+        public ExternalCommandSpec spec() {
+            return spec;
+        }
+
+        public void issue() {
+            String command = JsonUtil.toJsonString(this);
+            log.info("{}: Send out command: {} ", id, command);
+            controlChannel.println(command);
+        }
+    }
+
+    public class ErrorUpdater implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8))) {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    log.error("{}: (stderr of the process):{}", id, line);
+                    errMsg = line;
```

Comment:
It seems a little harsh to set a fatal error condition because the process logged something on stderr.  A lot of applications use stderr as an output.  Maybe we should just log this with `log.error`.

Reply:
Sure, done.